livekit-agents/livekit/agents/llm/chat_context.py

```diff
@@ -190,3 +190,37 @@ def from_dict(cls, data: dict) -> ChatContext:
         item_adapter = TypeAdapter(list[ChatItem])
         items = item_adapter.validate_python(data["items"])
         return cls(items)
+
+
+class _ReadOnlyChatContext(ChatContext):
+    """A read-only wrapper for ChatContext that prevents modifications."""
+
+    error_msg = (
+        "This is a read-only reference to the chat context. "
+        "Please use .copy() method to make a mutable copy."
+    )
+
+    class _ImmutableList(list):
+        def _raise_error(self, *args, **kwargs):
+            raise RuntimeError(_ReadOnlyChatContext.error_msg)
+
+        # override all mutating methods to raise errors
+        append = extend = pop = remove = clear = sort = reverse = _raise_error
+        __setitem__ = __delitem__ = __iadd__ = __imul__ = _raise_error
+
+        def copy(self):
+            return list(self)
+
+    @classmethod
+    def from_chat_context(cls, chat_ctx: ChatContext) -> _ReadOnlyChatContext:
+        return cls(chat_ctx.items)
+
+    def add_message(self, **kwargs) -> ChatMessage:
+        raise RuntimeError(_ReadOnlyChatContext.error_msg)
+
+    @property
+    def items(self) -> list[ChatItem]:
+        return self._ImmutableList(self._items)
+
+    def copy(self) -> ChatContext:
+        return ChatContext(self._items.copy())
```

Comment:
You don't even need to add those methods, It'll already raise? since items is _ImmutableList? We should maybe remove from_chat_context and use _ImmutableList inside the constructor?